src/Type/Definition/QueryPlan.php

```diff
@@ -0,0 +1,246 @@
+<?php
+
+declare(strict_types=1);
+
+namespace GraphQL\Type\Definition;
+
+use GraphQL\Error\Error;
+use GraphQL\Executor\Values;
+use GraphQL\Language\AST\FieldNode;
+use GraphQL\Language\AST\FragmentDefinitionNode;
+use GraphQL\Language\AST\FragmentSpreadNode;
+use GraphQL\Language\AST\InlineFragmentNode;
+use GraphQL\Language\AST\SelectionSetNode;
+use GraphQL\Type\Schema;
+use function array_filter;
+use function array_key_exists;
+use function array_keys;
+use function array_merge;
+use function array_merge_recursive;
+use function array_unique;
+use function array_values;
+use function count;
+use function in_array;
+use function is_array;
+use function is_numeric;
+
+class QueryPlan
+{
+    /** @var string[][] */
+    private $types = [];
+
+    /** @var Schema */
+    private $schema;
+
+    /** @var mixed[] */
+    private $queryPlan = [];
+
+    /** @var mixed[] */
+    private $variableValues;
+
+    /** @var FragmentDefinitionNode[]|null */
+    private $fragments;
+
+    /**
+     * @param FieldNode[]|null         $fieldNodes
+     * @param mixed[]                  $variableValues
+     * @param FragmentDefinitionNode[] $fragments
+     *
+     * @throws Error
+     */
+    public function __construct(ObjectType $parentType, Schema $schema, $fieldNodes, array $variableValues, array $fragments)
+    {
+        $this->schema         = $schema;
+        $this->variableValues = $variableValues;
+        $this->fragments      = $fragments;
+        $this->analyzeQueryPlan($parentType, $fieldNodes);
+    }
+
+    /**
+     * @return mixed[]
+     */
+    public function queryPlan() : array
+    {
+        return $this->queryPlan;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function referencedTypes() : array
+    {
+        return array_keys($this->types);
+    }
+
+    public function hasType(string $type) : bool
+    {
+        return count(array_filter($this->referencedTypes(), static function ($element) use ($type) {
```

Comment:
I guess we can typehint `string $element`. And then change variable name to `$referencedType`